### DISCUSSION TITLE
What is making filepaths relative in grep output?

### DISCUSSION THREAD

**Discussion by meedstrom** [minad/consult#1073] [Q&A]
Hello, I'm writing a command that greps in several directories (set in a variable).  I'm trying to track down what is causing filenames to print as relative paths in the `consult-grep` results.  I'd like full paths.

Here's a test command

```elisp
(defun look-many-dirs ()
  (interactive)
  (consult--grep "Grep in several hardcoded directories"
                 #'consult--grep-make-builder
                 '("~/.emacs.d/" "~/Documents" "~/Downloads/")
                 nil))
```

At first I used `#'consult--ripgrep-make-builder` and thought it was ripgrep responsible for the file path format, but no, ripgrep and vanilla grep will return full paths.  It's something either in consult,  or in the underlying Emacs completion systems.

I tracked down that the filename is stored in `'prefix-group` text property of each match, and this is fetched by the function `consult--prefix-group`, which doesn't transform the filename but gives it verbatim to `complete-with-action`.  So I am a bit lost as to where it is happening.

**Comment by minad** [comment]
> At first I used #'consult--ripgrep-make-builder and thought it was ripgrep responsible for the file path format, but no, ripgrep and vanilla grep will return full paths.

Actually `ripgrep` and `grep` return paths relative to the current directory, i.e., the `default-directory` where the process is started.

**Comment by meedstrom** [comment]
Not always.  I just tested `rg`, and it does that only if you provide a relative path (or no path).

What I'm doing is make a grep command for all org files, in various predetermined directories on the filesystem, so I'm providing full paths.  Then the results are full paths.  

But in consult--grep it's still shown as relative. Often with very long "../../../../../../.." strings. And again, I don't know if it's consult doing that conversion (guess not) or some underlying stuff.

**Reply by minad** [reply]
Yes, right. Consult first makes the input paths relative to the working directory in `consult--directory-prompt` and then passes those to `grep`. As a result `grep` returns relative paths.

**Reply by meedstrom** [reply]
Ah! Thanks. I'll try to hack that for my purposes.

**Reply by minad** [reply]
I haven't tested, but maybe the easiest solution is to just let-bind `default-directory`:

~~~ elisp
(let ((default-directory "/"))
  (consult--grep "Grep in several hardcoded directories"
                 #'consult--grep-make-builder
                 '("~/.emacs.d/" "~/Documents" "~/Downloads/")
                 nil))
~~~

**Reply by meedstrom** [reply]
Ah, I was about to use `cl-letf` to override `file-relative-name` with `identity`, but setting default-directory is less invasive.  It works, though it shows the paths with the first slash omitted, so they look a little unexpected.

Actually it might be nice to override `file-relative-name` with `consult--fast-abbreviate-file-name` ... but I get some messages `rg: ~/.emacs.d/: No such file or directory (os error 2)`.  Hmm.  Can't really expect good outcomes from overriding such a low level function though.

EDIT: To clarify:

```elisp
(cl-letf (((symbol-function #'file-relative-name) #'consult--fast-abbreviate-file-name))
 (consult--grep "Grep in several hardcoded directories"
                 #'consult--grep-make-builder
                 '("~/.emacs.d/" "~/Documents" "~/Downloads/")
                 nil)))
```

**Comment by meedstrom** [comment]
OK, that "No such file or directory" isn't about the override.  It occurs if I just redefine `consult--directory-prompt` so that the "ps" clause looks like `(ps (setq paths ps) default-directory)`.

Will probably keep looking into this later. At least I have a starting point.

EDIT: In the meantime, I do get good full paths with

```elisp
(cl-letf (((symbol-function #'file-relative-name) 
           (lambda (name &optional _dir) name)))
 (consult--grep "Grep in several hardcoded directories"
                 #'consult--grep-make-builder
                 '("~/.emacs.d/" "~/Documents" "~/Downloads/")
                 nil)))
```

so that is solved.  Abbreviating them is just an extra.
